src/core/inductive.ml

```diff
@@ -0,0 +1,263 @@
+  let case_of : sym -> tbox = fun scons ->
+    let rec case : tbox list -> term-> tbox = fun xs a ->
+      match Basics.get_args a with
+      | (Symb(s), l) ->
+          if s == sind then
+            let l = List.map lift l in
+            match List.rev l with
+            | []   -> prf_of_p (app (_Symb scons) (List.rev xs))
+            | t::q ->
+                let l = app t q in
+                let tmp t =
+                  _Appl (_Symb c.symb_prf) (_Appl (_Appl (_Vari p) l) t)
+                in
+                tmp (app (_Symb scons) (List.rev xs))
+          else fatal pos "%a is not a constructor of %a"
+                 pp_symbol scons pp_symbol sind
+      | (Prod(a,b), _) ->
+          let (x,b) =
+            if Bindlib.binder_occur b then
+              Bindlib.unbind b
+            else
+              let x = Bindlib.new_var mkfree "x" in
+              (x, Bindlib.subst b (Vari x))
+          in
+          let b = case ((Bindlib.box_var x)::xs) b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), l) ->
+                let b =
+                  if s == sind then
+                    let l = List.map lift l in
+                    match List.rev l with
+                    | []   -> _Impl (prf_of_p (Bindlib.box_var x)) b
+                    | _::_ ->
+                        let l = List.fold_left _Appl (_Vari p) l in
+                        _Impl (_Appl (_Symb c.symb_prf)
+                                 (_Appl l (Bindlib.box_var x)))
+                          b
+                  else b
+                in
+                if !log_enabled then
+                  log_ind "The current type : %a" Print.pp_term a;
+                let tmp = _Prod (Bindlib.box a) (Bindlib.bind_var x b) in
+                if !log_enabled then
+                  log_ind "The clause of the constructor [%a] is %a"
+                    Print.pp_term (Symb(scons))
+                    Print.pp_term (Bindlib.unbox tmp); tmp
+            | _ -> fatal pos "The type of %a is not supported"
+                     pp_symbol scons
+          end
+      | _ ->
+          fatal pos "The type of %a is not supported"
+            pp_symbol scons
+    in
+    case [] !(scons.sym_type)
+  in
+
+  let create_end symbol =
+    let rec aux t acc_app = match t with
+      | Type      ->
+          let x = Bindlib.new_var mkfree "x" in
+          let tmp =
+            _Prod symbol
+              (Bindlib.bind_var x (_Appl (_Symb c.symb_prf)
+                                     (_Appl acc_app (_Vari x))))
+          in
+          if !log_enabled then
+            log_ind "The end of the PI (2) : %a"
+              Print.pp_term (Bindlib.unbox tmp); tmp
+      | Prod(a,b) ->
+          let (x,b) =
+            if Bindlib.binder_occur b then
+              Bindlib.unbind b
+            else
+              let x = Bindlib.new_var mkfree "x" in
+              (x, Bindlib.subst b (Vari x))
+          in
```

Comment:
Thanks to factorize this function as it is used several times.

Reply:
Good idea! :)